feat: add endpoints to allow subscriptions management by user

wa0x6e opened the pull request:
Fix #46 and #57

Allow subscriber to manage their own subscriptions 

# Features

- Add a new `subscriptions` column in the table, to hold the subscriber's subscriptions.  Subscriptions is a list of email type (templateID). See schema.sql
- By default, on create, a subscriber will be subscribed to all email type
- When unsubscribing via the link in the email's footer, the unsubscribe page (envelop-ui) will show a form with a all subscriptions as checkbox, inviting user to select only the type of email he want to receive. By unchecking all options, he will be unsubscribed from all (deleted from db)
- Subscriber can also manage his subscription via his user profile. Same form with checkboxes, but on submit, a message signature will be required to verify wallet ownership.

This PR adds 2 new endpoints:

- `/subscriptionsList`: return a static list of all available email type, used by front end to show the form
- `/subscriber`: return an address' subscriptions 

# How to test 

Most of the changes are on the API side, and made for interaction with UI.
UI will be required for testing:

### testing from envelop-ui, unsubscribe page 

- require this branch https://github.com/snapshot-labs/envelop-ui/pull/3 , for UI interaction
- Boot an instance of envelop-ui, with the branch above
- Boot an instance of snapshot.org, with the branch https://github.com/snapshot-labs/snapshot/pull/3754
- Configure snapshot to use your local instance of envelop in the .env
- Subscribe to email notification via the snapshot UI
- You will receive a verification email
- Verify your email
- With the CLI scripts in envelop repo, send yourself a summary email
- In the email, go to footer, and click on the unsubscribe link
- The unsubscribe page will contains a form
- Submit with various options, and observe the changes in the database (column `subscriptions` should update)
- By unchecking all options and submit, the email should be removed from the database


### testing from snapshop ui, user profile email subscription section 

- require https://github.com/snapshot-labs/snapshot/issues/3886
- Click on your account in the page headr
- in the dropdown, click on email notifications
- in the modal, play with the form, and submit
- Sign the message
- observe changes in the database
-

wa0x6e commented:
This PR enables subscriptions management from envleop-ui

![Screenshot 2023-05-11 at 10 48 16 PM](https://github.com/snapshot-labs/envelop/assets/495709/ab29f36d-4441-454e-8163-6d69fdb59e8d)

wa0x6e commented:
Requesting @bonustrack review, as this PR requires database update

bonustrack commented:
@wa0x6e I've moved the server to the new DO account and added the "subscriptions" column on the table. But the deployment is failing with this error:
> ! Node v18+ is not supported on App Platform. Please update your package.json: https://do.co/apps-buildpack-node

wa0x6e commented:
> @wa0x6e I've moved the server to the new DO account and added the "subscriptions" column on the table. But the deployment is failing with this error:
> 
> > ! Node v18+ is not supported on App Platform. Please update your package.json: https://do.co/apps-buildpack-node
 
I'll take a look, may have to use dockerfile for deployment.

Todmy commented:
During subsequent try to subscribe to an already subscribed email, we do nothing in DB (we have the primary key for the combination of address and email), but we send an email notification to the user with a request to follow a link for subscription confirmation. Don't you think that we shouldn't send this email, since we just spend company money? 
Even more, by creating a simple script some attackers can create unpleasant expenses that we have to pay to SendGrid.

wa0x6e commented:
> During subsequent try to subscribe to an already subscribed email, we do nothing in DB (we have the primary key for the combination of address and email), but we send an email notification to the user with a request to follow a link for subscription confirmation. Don't you think that we shouldn't send this email, since we just spend company money? Even more, by creating a simple script some attackers can create unpleasant expenses that we have to pay to SendGrid.

You're right. On subscribe, we should send the email if:
- the email/address does not exist
- the email/address exist, but not verified

All other cases should skip the email sending.

To avoid bot and signature reuse, maybe we should add an additional salt params to the request. Feel free to open a PR

Todmy commented:
I have noticed that on FE we sign the message and send this signature to BE, but during the subscription, we don't use it for validation of sender address ownership.

Todmy commented:
after subscription, a user has no default `subscriptions` in DB. I believe we should add some defaults, don't we? Otherwise, he/she will never get any email from us if he/she doesn't edit subscription preferences.

Todmy commented on src/helpers/utils.ts:
```diff
@@ -35,8 +42,32 @@ export async function verify(email: string, address: string) {
+export async function update(email: string, address: string, subscriptions: string[]) {
+  const subs = sanitizeSubscriptions(subscriptions);
+  if (subs.length === 0) {
+    return await db.queryAsync('DELETE FROM subscribers WHERE email = ? and address = ? LIMIT 1', [
```
another problem with completely removing email is that if a user accidentally unsubscribed from all subscription, and then want to subscribe again, he needs to pass the whole subscription flow, which is not user-friendly.

wa0x6e replied:
This is by design, see discussion here: https://github.com/snapshot-labs/envelop/pull/28#issuecomment-1475966415

Todmy replied:
I believe the `update` function should be used for subscription management in the snapshot portal,  If we are talking about the `unsubscribe` function, which can be used only inside the subscription inside the email, then probably it can remove the email completely.
If we leave it as it is, the case I have described in the second comment should be handled on the snapshot portal.

Todmy replied:
@bonustrack what do you think about it?

wa0x6e replied:
I am lost, what the request/issue ?

Todmy replied:
My proposal is not to remove the email completely from the table during the `update` flow but remove all subscriptions instead. But the `unsubscribe` flow can remain as you did - it will remove email completely from the list.

wa0x6e replied:
I am also in favor of keeping the user email on unsubscription, specially with the new preferences form: https://user-images.githubusercontent.com/7693369/238181550-d088db3a-d03a-4607-a3ea-6f57720252ef.png

By unchecking all, user may think that he can come back later to check some again, but it will not be the case.

Todmy replied:
resolve and move discussions to https://github.com/snapshot-labs/envelop/issues/62

wa0x6e commented on src/sign/index.ts:
```diff
@@ -1,31 +1,30 @@
-const NAME = 'envelop';
-const VERSION = '1';
+const NAME = 'snapshot';
+const VERSION = '0.1.4';
```
version should be fixed, it has to have the same value as the one defined in the `sign` function on snapshot UI side.

Todmy replied:
ok, it's not critical

wa0x6e commented on src/rpc.ts:
```diff
@@ -45,16 +48,54 @@ router.post('/', async (req, res) => {
       return rpcError(res, 500, 'Unable to authenticate your verification link', id);
     } else if (method === 'snapshot.unsubscribe') {
       if (verifyUnsubscribe(params.email, params.signature)) {
```
The signer for `verifyUnsubscribe` is envelop backend itself, and thus is `WALLET_PRIVATE_KEY`

Todmy replied:
Everything interconnected.
I think for unsubscribing inside the email you need to create a link with the hash sum of the client email and sult. We can assume that only the user can have access to the email and that's why only the user can click this link.
But as an option the link could lead to `envelop-ui` and the user should sign it as well to confirm his action.
The first approach is simpler, and the second, I think, is more secure.

Todmy replied:
yes, get you

wa0x6e commented on src/helpers/utils.ts:
```diff
@@ -35,8 +42,32 @@ export async function verify(email: string, address: string) {
+  }
+}
+
+export async function unsubscribe(email: string, subscriptions?: string[]) {
```
`unsubscribe` does not ask for a signature from user. This is made for the unsubscribe link in the email footer

Todmy replied:
I got the use case, thank you.
Unfortunately, it asks: https://github.com/snapshot-labs/envelop/pull/55/files#diff-6b984d0f3bee25da0d949d2e9fddd45f75509458a9b0c05b589af909ffe0911aR50
w without signature `verifyUnsubscribe` will fail. 
If I get you right, this flow is for unsubscribing from email. In this case, you need to create something similar to `validate email` flow

Todmy replied:
this case is related to all our discussions about unsubscribing. I initially missed the point of this flow. Now I got it.

wa0x6e commented on src/rpc.ts:
```diff
@@ -45,16 +48,54 @@ router.post('/', async (req, res) => {
     } else if (method === 'snapshot.unsubscribe') {
       if (verifyUnsubscribe(params.email, params.signature)) {
-        await unsubscribe(params.email);
+        await unsubscribe(params.email, params.subscriptions);
```
`verifyUnsubscribe` is here to prevent that, you can not pass any email, else the signature will fail

Todmy replied:
with the current implementation, no one can unsubscribe, because we assume that the signer is the one whose private key is defined in `process.env.WALLET_PRIVATE_KEY`. That's why I proposed avoiding `signer || wallet.address`. 
Looking at `verifyUnsubscribe` we can assume that it receives only email and signature, from the signature you can get the signer, but you cannot compare it with the signer, since we don't pass it to `verifyUnsubscribe`.

wa0x6e replied:
backend sign(email) then send signature to user
user return the signature, backend only verify(email), and check that signer is himself

Todmy replied:
Thank you I got the case.

wa0x6e commented on src/helpers/utils.ts:
```diff
@@ -45,8 +76,18 @@ export async function getEmailAddresses(email: string) {
 
-export async function getUniqueEmails() {
-  return await db.queryAsync('SELECT DISTINCT email FROM subscribers WHERE verified > 0');
+export async function getUniqueEmails(subscriptionType: string) {
+  return await db.queryAsync(
+    `SELECT DISTINCT email FROM subscribers WHERE verified > 0 AND JSON_CONTAINS(subscriptions, '${JSON.stringify(
+      sanitizeSubscriptions(subscriptionType)[0]
+    )}')`
+  );
+}
```
Aren't you missing the `JSON.stringify` ?

Todmy replied:
it's redundant for this case, but we can add it. Since by stringifying string, you receive the string with quotes: 
```js
const str = 'test';
console.log(JSON.stringify(str) === '"test"'); // true
```
if you think that it should be added, then:
```suggestion
export async function getUniqueEmails(subscriptionType: string) {
  const subscription = sanitizeSubscriptions(subscriptionType)[0];
  const stringifiedSubscription = JSON.stringify(subscription);
  
  return await db.queryAsync(
    `SELECT DISTINCT email FROM subscribers WHERE verified > 0 AND JSON_CONTAINS(subscriptions, '${stringifiedSubscription}')`
  );
}
```

wa0x6e replied:
yes, we do need both the single and double here

wa0x6e commented on src/rpc.ts:
```diff
@@ -45,16 +48,54 @@ router.post('/', async (req, res) => {
+
+  try {
+    const subscribers = await getAddressSubscriptions(address);
+    if (!subscribers || !subscribers[0]) {
```
1 wallet can be attached to only on email

Todmy replied:
than something wrong, because I have succeeded to register with multiple emails:
 
<img width="491" alt="Screenshot 2023-05-12 at 19 58 25" src="https://github.com/snapshot-labs/envelop/assets/7693369/2c5efa7e-3d70-4a6b-b324-1ad0c2b60bfa">

wa0x6e replied:
This is expecting that the PR https://github.com/snapshot-labs/envelop/pull/52 was pulled, which add a unique index on these columns preventing duplicate.

By the way, without the new unique index, isn't the primary key supposed to prevent that too ?

Todmy replied:
yes, but in PR #52 you have added a `UNIQUE` index, but this will not fix the case. Because you still will be able to attach different emails to the same address. I have added the comment to #52 
If you want to prevent this, you need to add `UNIQUE` index to the address column only.

wa0x6e replied:
You are correct, I'll add a check

wa0x6e replied:
> what if the client subscribed to multiple emails? Maybe we should handle it on the step of subscription and notify him/her that he already subscribed with another email?

He can subscribe to many emails, but he can only verify only one. All subsequent verification should fail, since one address can only be associated to a verified email.

Todmy replied:
Checked. Looks like it is possible. I verified multiple emails:

<img width="847" alt="Screenshot 2023-05-12 at 21 28 28" src="https://github.com/snapshot-labs/envelop/assets/7693369/3f81a62d-5333-4f65-9e9f-a8f0e65a7e84">

btw, I don't use UI, I call all endpoints with curl

wa0x6e replied:
WIP, will prevent verifying multiple emails on the same address

wa0x6e replied:
Done, only a single email can be verified per wallet

wa0x6e replied:
Additionally, you can also check that from UI side, by querying /subscriber with the wallet address

wa0x6e commented on src/helpers/utils.ts:
```diff
@@ -21,6 +22,12 @@ export function rpcError(res: Response, code: number, e: unknown, id: string | n
+    SUBSCRIPTIONS_TYPE.includes(item)
+  ) as typeof SUBSCRIPTIONS_TYPE;
+}
+
```
Let's just use string | string[], these are user submitted values, we're not sure they are templateId, the point of this function is to make sure they are templateId

wa0x6e commented on src/helpers/utils.ts:
```diff
@@ -35,8 +42,32 @@ export async function verify(email: string, address: string) {
+    ]);
+  } else {
+    return await db.queryAsync(
+      'UPDATE subscribers SET subscriptions = ? WHERE email = ? AND address = ? LIMIT 1',
```
A convention from my side to know at first glance that this is only updating a single row, regardless of what the schema says. Should we remove it ?

Todmy replied:
I got you. Not critical. Up to you to remove or leave it

Todmy commented on src/rpc.ts:
```diff
@@ -45,16 +48,54 @@ router.post('/', async (req, res) => {
+  }
+});
+
+router.get('/subscriptionsList', (req, res) => {
```
At FE will be used only keys that will be mapped to translations. Maybe we can simplify this endpoint(just returning an array of keys)?

wa0x6e replied:
You already got the keys in the objects if you need. The full object is used by envelop-ui, which does not have i18n

wa0x6e commented:
> after subscription, a user has no default subscriptions in DB. I believe we should add some defaults, don't we? Otherwise, he/she will never get any email from us if he/she doesn't edit subscription preferences.

I think we should transform this into a feature, and make it so that empty subscriptions == all subscriptions. That way, we can add new subscription type later, without updating all rows in the db.

wa0x6e commented:
> > after subscription, a user has no default subscriptions in DB. I believe we should add some defaults, don't we? Otherwise, he/she will never get any email from us if he/she doesn't edit subscription preferences.
> 
> I think we should transform this into a feature, and make it so that empty subscriptions == all subscriptions. That way, we can add new subscription type later, without updating all rows in the db.

Added

Todmy commented on src/helpers/utils.ts:
```diff
@@ -9,34 +16,91 @@ export function rpcSuccess(res: Response, result: string, id: string | number) {
+    throw new Error('ADDRESS_ALREADY_VERIFIED_WITH_ANOTHER_EMAIL');
+  }
+
+  const a = (await db.queryAsync(
```
can it be a more descriptive name? something like `dbResponse`?

Todmy reviewed (changes requested):
need to remove `email` from the body of `snapshot.update` JSON-RPC method

Todmy commented on src/helpers/utils.ts:
```diff
@@ -76,16 +76,26 @@ export async function verify(email: string, address: string) {
+      )} AND verified > 0`,
+      [JSON.stringify(sanitizeSubscriptions(subscriptions)), ...Object.values(fields)]
     );
   }
```
can we do it a little bit more readable?
```ts
export async function update(email: string, address: string, subscriptions: string[]) {
  const fields: Record<string, string> = {};
  if (email.length > 0) {
    fields['email = ?'] = email;
  }
  if (address.length > 0) {
    fields['address = ?'] = address;
  }

  const subs = sanitizeSubscriptions(subscriptions);
  const whereQueryChunk = Object.keys(fields).join(' AND ');

  if (subs.length === 0) {
    return db.queryAsync(
      `DELETE FROM subscribers WHERE ${whereQueryChunk}`,
      Object.values(fields)
    );
  } else {
    const stringifiedSubs = JSON.stringify(subs)
    return db.queryAsync(
      `UPDATE subscribers SET subscriptions = ? WHERE ${whereQueryChunk} AND verified > 0`,
      [stringifiedSubs, ...Object.values(fields)]
    );
  }
}
```

Todmy commented on src/sign/index.ts:
```diff
@@ -42,18 +52,23 @@ export function verifySubscribe(email: string, address: string, signature: strin
+  );
+}
+
+export function verifyVerify(email: string, address: string, signature: string) {
```
created #91. Comment out of scope

Todmy commented:
> `snapshot.update` now does not require `email`.
> 
> It either accept `email`, `address`, or both.

Sorry, but it doesn't work without email. Probably because of this [line](https://github.com/snapshot-labs/envelop/blob/74285e1626a91e6976af216d3700c6ef1ef3f5de/src/rpc.ts#L63)

Todmy commented:
> > > after subscription, a user has no default subscriptions in DB. I believe we should add some defaults, don't we? Otherwise, he/she will never get any email from us if he/she doesn't edit subscription preferences.
> > 
> > 
> > I think we should transform this into a feature, and make it so that empty subscriptions == all subscriptions. That way, we can add new subscription type later, without updating all rows in the db.
> 
> Added

sorry, I don't see changes. Is it in another PR?
